plugin-commons/src/main/java/org/sonarsource/sonarlint/core/plugin/commons/container/SpringComponentContainer.java

```diff
@@ -0,0 +1,259 @@
+   */
+  @Override
+  public Container add(Object... objects) {
+    for (Object o : objects) {
+      if (o instanceof Class) {
+        Class<?> clazz = (Class<?>) o;
+        context.registerBean(componentKeys.ofClass(clazz), clazz);
+        declareProperties(o);
+      } else if (o instanceof Iterable) {
+        ((Iterable<?>) o).forEach(this::add);
+      } else {
+        registerInstance(o);
+        declareProperties(o);
+      }
+    }
+    return this;
+  }
+
+  private <T> void registerInstance(T instance) {
+    Supplier<T> supplier = () -> instance;
+    Class<T> clazz = (Class<T>) instance.getClass();
+    context.registerBean(componentKeys.ofInstance(instance), clazz, supplier);
+  }
+
+  /**
+   * Extensions are usually added by plugins and we assume they don't support any injection-related annotations.
+   * Spring contexts supporting annotations will fail if multiple constructors are present without any annotations indicating which one to use for injection.
+   * For that reason, we need to create the beans ourselves, using ClassDerivedBeanDefinition, which will declare that all constructors can be used for injection.
+   */
+  private Container addExtension(Object o) {
+    if (o instanceof Class) {
+      Class<?> clazz = (Class<?>) o;
+      var bd = new ClassDerivedBeanDefinition(clazz);
+      context.registerBeanDefinition(componentKeys.ofClass(clazz), bd);
+    } else if (o instanceof Iterable) {
+      ((Iterable<?>) o).forEach(this::addExtension);
+    } else {
+      registerInstance(o);
+    }
+    return this;
+  }
+
+  @Override
+  public <T> T getComponentByType(Class<T> type) {
+    try {
+      return context.getBean(type);
+    } catch (Exception t) {
+      throw new IllegalStateException("Unable to load component " + type, t);
+    }
+  }
+
+  @Override
+  public <T> Optional<T> getOptionalComponentByType(Class<T> type) {
+    try {
+      return Optional.of(context.getBean(type));
+    } catch (NoSuchBeanDefinitionException t) {
+      return Optional.empty();
+    }
+  }
+
+  @Override
+  public <T> List<T> getComponentsByType(Class<T> type) {
+    try {
+      return new ArrayList<>(context.getBeansOfType(type).values());
+    } catch (Exception t) {
+      throw new IllegalStateException("Unable to load components " + type, t);
+    }
+  }
+
+  public AnnotationConfigApplicationContext context() {
+    return context;
+  }
+
+  public void execute() {
+    RuntimeException r = null;
+    try {
+      startComponents();
+    } catch (RuntimeException e) {
+      r = e;
+    } finally {
+      try {
+        stopComponents();
+      } catch (RuntimeException e) {
+        if (r == null) {
+          r = e;
+        }
+      }
+    }
+    if (r != null) {
+      throw r;
+    }
+  }
+
+  @Override
+  public SpringComponentContainer startComponents() {
+    doBeforeStart();
+    context.refresh();
+    doAfterStart();
```

Comment:
As discussed this morning, I think it would simplify the code if we get all the Startables in the container at this point and start them. We could get rid of `SpringInitStrategy` and `LazyUnlessStartableStrategy`, and part of the `StartableBeanPostProcessor` would move here. Maybe a feedback to make to the SQ team ?

Reply:
One problem I see is that we should start in the dependency order. So the way it is implemented today is better IMO.